third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1186 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Overview of approach:
+// - We visit function calling inline functions reachable from compilation targets in a bottom-up fashion,
+//   storing rewritten functions in a map to simplify further processing.
+//   - Change to the program happens at the end.
+//
+// - struct `Inliner`
+//   - holds the map recording inlined function bodies so that we don't need to modify the program until the end.
+//   - `try_inlining_in` function is the entry point for each function needing inlining.
+//
+// - struct `OutlerInlinerRewriter` uses trait `ExpRewriterFunctions` to rewrite each call in the target.
+//   - `rewrite_call` recognizes inline functions and rewrites them using `InlinedRewriter::inline_call`
+//
+// - struct `InlinedRewriter` uses trait `ExpRewriterFunctions` to rewrite the inlined function body.
+//   - `rewrite_exp` disallows Return expressions
+//
+//   - `rewrite_enter_scope` and `rewrite_exit_scope` record lambda free vars which are shadowed by
+//      a local scope, so that uses can be renamed to shadow vars appropriately.
+//
+//   - `rewrite_node_id` instantiates type_args on every node in the inlined function
+//
+//   - `rewrite_local_var` replaces symbol uses that are shadowed with the shadow symbol.
+//   - `rewrite_temporary` replaces references to function parameters with parameter symbols, also shadowing as needed.
+//
+//   - `rewrite_invoke` handles calls to lambda parameters within the inlined function
+//      - Note that lambda bodies are not rewritten during inlining, but are kept intact
+//
+//   - `rewrite_pattern` replaces syms in a pattern with shadow symbols as necessary
+//
+// - struct `InlinedRewriter` also has various methods to support rewriting a call
+//   - `inline_call` is the entry point for rewriting a call to an inline function.
+//
+//   - `create_shadow_symbols` creates a shadow `Symbol` for each lambda free variable to use if
+//      a local variable conflicts with that variable.  It uses `SymbolPool::shadow` to create
+//      shadow symbols which print out the same but don't conflict.
+//
+//   - `get_shadow_symbol` checks whether a `Symbol` conflicts with lambda free variables, and if so,
+//     returns the shadow symbol to use instead..
+//
+//   - `parameter_list_to_pattern` is a helper to convert a list of `Parameter` into a `Pattern`,
+//      suitable for use in the body.
+//
+//   - `construct_inlined_call_expression` is a helper to build the expression corresponding to
+//      { let params=actuals; body } used for both lambda inlining and inline function inlining.
+//
+//   - rewrite_pattern_vector is a helper for `rewrite_pattern`
+//
+
+// TODO:
```

Comment:
I think we are missing out on visibility checks for inline functions. We should discuss on how to handle this.

Reply:
I have a draft PR waiting for this one to land at brmaptos:brm-issue-10949.